bd-artifact-upload/src/uploader.rs

```diff
@@ -0,0 +1,666 @@
+use bd_proto::protos::client::artifact::ArtifactUploadIndex;
+use bd_proto::protos::client::artifact::artifact_upload_index::Artifact;
+use bd_proto::protos::logging::payload::data::Data_type;
+use bd_proto::protos::logging::payload::{BinaryData, Data};
+use bd_runtime::runtime::{ConfigLoader, IntWatch, artifact_upload};
+use bd_shutdown::ComponentShutdown;
+use bd_time::{OffsetDateTimeExt, TimeProvider};
+use mockall::automock;
+use std::collections::{HashSet, VecDeque};
+use std::path::PathBuf;
+use std::sync::{Arc, LazyLock};
+#[cfg(test)]
+use tests::TestHooks;
+use uuid::Uuid;
+
+/// Root directory for all files used for storage and uploading.
+pub static REPORT_DIRECTORY: LazyLock<PathBuf> = LazyLock::new(|| "report_uploads".into());
+
+/// The index file used for tracking all of the individual files.
+pub static REPORT_INDEX_FILE: LazyLock<PathBuf> = LazyLock::new(|| "report_index.pb".into());
+
+//
+// NewUpload
+//
+
+// TODO(snowp): Consider allowing passing an open file handle instead of having to hold the data in
+// memory while entry is pending within the channel.
+#[derive(Debug)]
+struct NewUpload {
+  uuid: Uuid,
+  contents: Vec<u8>,
+  state: LogFields,
+}
+
+// Used for bounded_buffer logs
+impl std::fmt::Display for NewUpload {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    write!(
+      f,
+      "NewUpload {{ uuid: {}, contents: {} }}",
+      self.uuid,
+      self.contents.len()
+    )
+  }
+}
+
+impl MemorySized for NewUpload {
+  fn size(&self) -> usize {
+    size_of::<Uuid>() + size_of::<Vec<u8>>() + self.contents.capacity() * size_of::<u8>()
+  }
+}
+
+//
+// Stats
+//
+
+struct Stats {
+  uploaded: Counter,
+
+  // TODO(snowp): For now we just emit metrics on drops but we probably want a more robust
+  // mechanism for keeping track of the data that we're dropping either due to overflows or intent
+  // rejections.
+  dropped: Counter,
+  dropped_intent: Counter,
+  accepted_intent: Counter,
+}
+
+impl Stats {
+  fn new(scope: &Scope) -> Self {
+    Self {
+      uploaded: scope.counter("uploaded"),
+      dropped: scope.counter("dropped"),
+      dropped_intent: scope.counter("dropped_intent"),
+      accepted_intent: scope.counter("accepted_intent"),
+    }
+  }
+}
+
+#[automock]
+pub trait Client: Send + Sync {
+  fn enqueue_upload(&self, contents: Vec<u8>, state: LogFields) -> anyhow::Result<Uuid>;
+}
+
+pub struct UploadClient {
+  upload_tx: bd_bounded_buffer::Sender<NewUpload>,
+  counter_stats: SendCounters,
+}
+
+impl Client for UploadClient {
+  /// Dispatches a payload to be uploaded, returning the associated artifact UUID.
+  fn enqueue_upload(&self, contents: Vec<u8>, state: LogFields) -> anyhow::Result<Uuid> {
+    let uuid = uuid::Uuid::new_v4();
+
+    let result = self
+      .upload_tx
+      .try_send(NewUpload {
+        uuid,
+        contents,
+        state,
+      })
+      .inspect_err(|e| log::warn!("failed to enqueue artifact upload: {e:?}"));
```

Comment:
Related to my other comment, should this also increment the dropped metric? That would cover it.

Reply:
The bounded buffer comes with a bunch of stats already so I was planning on using that, including whether the drop was due to size or count